### ISSUE TITLE
What would be needed to train the model for other purposes?

### ISSUE THREAD

**Issue by munybt** [tech-srl/code2vec#42]
How could I adapt code2vec to train the model for something else than the method's name?
In my case, I want the prediction to be a boolean.

**Comment by urialon** [comment]
Hi,
You can feed input with any number of classes, including two.
See:
[https://github.com/tech-srl/code2vec#extending-to-other-languages](https://github.com/tech-srl/code2vec#extending-to-other-languages)
So instead of the method name, you can feed "true" and "false".

By the way,
[code2seq](https://github.com/tech-srl/code2seq/) has a better encoder than code2vec, and even though it can generate longer sequences, it can be used for binary classification in the same way.

**Comment by munybt** [comment]
So, all I have to do is modifying the Extractor to change the first token in each line?

I visited the code2seq page and I couldn't tell the difference from code2vec. Could you please clarify?

**Comment by urialon** [comment]
Exactly.

In general, the encoder of code2seq is better because it learns paths and names using LSTMs, node-by-node, while in code2vec the paths are monolithic.
For additional differences between code2vec and code2seq, see Section 6 in [the code2seq paper](https://openreview.net/pdf?id=H1gKYo09tX).

**Comment by DavidSUN2** [comment]
Hi @munybt @urialon,
I'm sorry though this has been closed.
If I train this model for binary classification problem with label 'true' and 'false', according to your discussion above, I only need to modify extractor file. I have 2 related question:
1. I have checked the model.py file. I found the following loss function:
```
            loss = tf.reduce_sum(tf.nn.sparse_softmax_cross_entropy_with_logits(
                labels=tf.reshape(words_input, [-1]),
                logits=logits)) / batch_size
```
do I need to change this part used in model.py file? Or any other part in model.py I need to modify for binary classification problem?
2. About extractor file, which specific part do I need to change? Could you please give me some hints?
Thanks in advance.

**Comment by urialon** [comment]
Hi @DavidSUN2 ,
I don't think you need to change the loss. Binary classification is a special case of multi-class classification. As long as the model "sees" only two possible labels during training - this is binary classification.
If you change the loss to `tf.nn.sigmoid_cross_entropy_with_logits` you can allow the model to assign multi labels, i.e, the same example can be both "true" and "false".

Regarding the extractor - it depends on where do you get your true/false labels from.
This is the line that prints the method name as the "label":
[https://github.com/tech-srl/code2vec/blob/master/JavaExtractor/JPredict/src/main/java/JavaExtractor/FeaturesEntities/ProgramFeatures.java#L21](https://github.com/tech-srl/code2vec/blob/master/JavaExtractor/JPredict/src/main/java/JavaExtractor/FeaturesEntities/ProgramFeatures.java#L21)

So instead of `stringBuilder.append(name)` you can `stringBuilder.append` anything else. I hope it helps.
